src/cpp/session/modules/SessionPackageProvidedExtension.cpp

```diff
@@ -0,0 +1,250 @@
+/*
+ * SessionPackageProvidedExtension.cpp
+ *
+ * Copyright (C) 2009-16 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <session/SessionPackageProvidedExtension.hpp>
+
+#include <boost/regex.hpp>
+#include <boost/foreach.hpp>
+#include <boost/date_time/posix_time/posix_time.hpp>
+
+#include <core/Algorithm.hpp>
+#include <core/FileSerializer.hpp>
+#include <core/text/DcfParser.hpp>
+
+#include <session/SessionModuleContext.hpp>
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace ppe {
+
+Error parseDcfResourceFile(
+      const FilePath& resourcePath,
+      boost::function<Error(const std::map<std::string, std::string>&)> callback)
+{
+   Error error;
+
+   // read dcf contents
+   std::string contents;
+   error = core::readStringFromFile(resourcePath, &contents, string_utils::LineEndingPosix);
+   if (error)
+      return error;
+
+   // attempt to parse as DCF -- multiple newlines used to separate records
+   boost::regex reSeparator("\\n{2,}");
+   boost::sregex_token_iterator it(contents.begin(), contents.end(), reSeparator, -1);
+   boost::sregex_token_iterator end;
+
+   for (; it != end; ++it)
+   {
+      // invoke parser on current record
+      std::map<std::string, std::string> fields;
+      std::string errorMessage;
+      error = text::parseDcfFile(*it, true, &fields, &errorMessage);
+      if (error)
+         return error;
+      
+      // invoke callback on parsed dcf fields
+      error = callback(fields);
+      if (error)
+         return error;
+   }
+   
+   return Success();
+}
+
+Indexer::Indexer() : index_(0), n_(0), running_(false) {}
+
+void Indexer::addWorker(boost::shared_ptr<Worker> pWorker)
+{
+   workers_.push_back(pWorker);
```

Comment:
Probably not, but clients should generally be registering their workers quite early in the startup process (before deferred init, and hence before the indexer ever gets a chance to run) so in practice that shouldn't be an issue.